src/pipdeptree/_discovery.py

```diff
@@ -1,35 +1,39 @@
 from __future__ import annotations
 
+import sys
+from importlib.metadata import distributions
+from pathlib import Path
 from typing import TYPE_CHECKING
 
 if TYPE_CHECKING:
-    from pip._vendor.pkg_resources import DistInfoDistribution
+    from importlib.metadata import Distribution
+
+
+def is_virtual_environment():
+    return hasattr(sys, "prefix") and sys.prefix != sys.base_prefix
+
+
+def get_site_packages_directory():
+    if is_virtual_environment():
+        return Path(sys.prefix) / "lib" / f"python{sys.version_info.major}.{sys.version_info.minor}" / "site-packages"
+    return None
 
 
 def get_installed_distributions(
     local_only: bool = False,  # noqa: FBT001, FBT002
     user_only: bool = False,  # noqa: FBT001, FBT002
-) -> list[DistInfoDistribution]:
-    try:
-        from pip._internal.metadata import pkg_resources  # noqa: PLC0415, PLC2701
-    except ImportError:
-        # For backward compatibility with python ver. 2.7 and pip
-        # version 20.3.4 (the latest pip version that works with python
-        # version 2.7)
-        from pip._internal.utils import misc  # noqa: PLC0415, PLC2701 # pragma: no cover
-
-        return misc.get_installed_distributions(  # type: ignore[no-any-return,attr-defined]
-            local_only=local_only,
-            user_only=user_only,
-        )
-
-    else:
-        dists = pkg_resources.Environment.from_paths(None).iter_installed_distributions(
-            local_only=local_only,
-            skip=(),
-            user_only=user_only,
-        )
-        return [d._dist for d in dists]  # type: ignore[attr-defined] # noqa: SLF001
+) -> list[Distribution]:
+    dists = distributions()
+
+    if local_only:
+        site_package = get_site_packages_directory()
+        dists = (d for d in dists if d.files and site_package in d.files)
+
+    if user_only:
+        user_site = Path(sys.modules["site"].getusersitepackages())
+        dists = (d for d in dists if d.files and user_site in d.files)
```

Comment:
Sadly, this won't work.

When `local_only` is set,  it seems you are creating a global site directory (FYI it would be better to use `site.getsitepackages()` instead of a hardcoded value),  are looking to see if the current distribution has files that are in the global site directory, and include it if it does (I think you meant to negate this). A better approach here may be to get the base virtual env path and use that instead to grab only the virtual-env-related packages.

For the `user_only` code path, it would be better to just import the `site` package and do `site.getusersitepackages()`.

Another problem I see is how you are comparing the file paths. You should check if the files share a common directory path to the virtual env/user site directories, not just perform an equality check like the following:
```python
# This will fail because the strings aren't equal
"/home/vscode/.local/lib/python3.8/site-packages" == "/home/vscode/.local/lib/python3.8/site-packages/pluggy-1.4.0.dist-info/INSTALLER"
```

And instead of using `Distribution.files`, you could instead do `Distribution.locate_file("")` as it seems to give the base directory path of where the distribution resides (see [here](https://github.com/python/cpython/blob/95c340a86b828cac6c0a2e4f2fa8a96695388c73/Lib/importlib/metadata.py#L493C1-L494C40)):
```console
>>> from importlib.metadata import distribution
>>> d = distribution("pipdeptree")
>>> d.locate_file("")
PosixPath('/home/vscode/.local/lib/python3.8/site-packages')
>>>
```

Reply:
> I rewrite the left part considering your advice, but I got confused at path comparing part mentioned. Aren't site.getusersitepackages() and str(d.locate_file("")) equal? If so, can I use in to check if dist.locate_file("") in site_packages, which means dist is in virtual env.

Yes, you are correct; it looks like I worded this suggestion kind of weird. Since `d.locate_file("")` just gives the path where the distribution resides, your way should work.